### ISSUE TITLE
Add card scanner

### ISSUE THREAD

**Issue by ertrzyiks** [ertrzyiks/woodtime#78]
✅ PRE-IMPLEMENTATION PLAN (React + Vite + OpenCV.js + TF.js)
Focus: Grid recognition + cell extraction + ML-ready data

1. Project Structure

Add a new route with a button to start recording and a preview from the camera

OpenCV.js is either:

loaded via public/opencv.js, or

injected via CDN in index.html

TF.js loaded via npm (@tensorflow/tfjs)

2. Grid Recognition Pipeline (Core Objective)
This step determines where the grid is on the physical card.
It must work regardless of rotation, tilt, distance, or lighting.
2.1 Input

A video frame from the phone camera (HTML video → canvas)

720p is enough

2.2 Step 1 — Preprocessing
Convert frame → usable CV image:
RGB → Grayscale → Gaussian Blur

Purpose:

remove noise

prepare for contour detection

2.3 Step 2 — Edge Detection
Use Canny:

highlights grid borders

highlights paper outline

We are looking for the outer border of the paper, not the grid lines inside.

2.4 Step 3 — Contour Detection
Find largest quadrilateral:
Algorithm:

cv.findContours()

filter contours by area

approximate polygon → cv.approxPolyDP

pick the largest 4-point shape

This gives the outline of the paper/card.
Why?
Because the grid is printed inside the card, not edge-to-edge.
The paper is the stable object.

2.5 Step 4 — Perspective Transform
Warp the detected quadrilateral to a perfect rectangle.

Output resolution = fixed, e.g. 800×600

This gives a consistent view regardless of camera angle.

This is the single highest-impact step for accuracy.

3. Recognizing the GRID inside the rectified card
Now you have a top-down, undistorted image of the card.
Grid detection must find the rows & columns, then extract each cell.
You have two possible strategies:

Strategy A — Known Fixed Grid (recommended if grid is printed consistently)
Assume:

fixed number of rows (e.g., 5)

fixed number of columns (e.g., 7)

fixed spacing

We simply:

measure width/height of warped image

divide by row/column counts

crop each cell

Pros:

dead simple

reliable

fast

ML model gets uniform samples

Cons:

requires consistent print design

This is almost always enough.

Strategy B — Line Detection (if print layout can vary)
Steps:

Run adaptive threshold

Use Hough Line Transform

Detect vertical & horizontal grid lines

Sort intersections → cell rectangles

Crop cells dynamically

Pros:

automatically extracts grid from any layout

Cons:

more complex

sensitive to weak lines

Most projects do not need this.

4. Cell Extraction for ML
For each cell rectangle:
4.1 Crop
Use:
const roi = warpedMat.roi(new cv.Rect(x, y, width, height));

4.2 Resize to ML input
32×32 or 64×64 grayscale recommended.
cv.resize(roi, roi, new cv.Size(32, 32));

4.3 Convert to tensor format
The ML model expects:
Float32Array or Uint8Array, normalized to [0, 1]:
pixel = value / 255
shape = [1, 32, 32, 1]

4.4 Output
For each cell → a ready-to-predict tensor.

5. Prediction (Model Already Trained — Option B)
Once cells are extracted:
const prediction = model.predict(cellTensor);
const [emptyProb, stampedProb] = prediction.dataSync();

Apply threshold (e.g., 0.6):
if (stampedProb > 0.6) cell = STAMPED;
else cell = EMPTY;

6. Stabilization Logic (Optional but Recommended)
To avoid flickering results:

collect predictions over last N frames (e.g., 5)

take median result per cell

Adds stability when the user’s hand is shaking.

7. Debugging Tools (React Components)
CameraView.jsx

Renders <video>

Manages camera stream

Triggers processing loop using requestAnimationFrame

DebugCanvas.jsx

Shows:

edges

warped card preview

cell bounding boxes

Helps tune thresholds

8. Assumptions for Implementation
(A) Printed card is fixed format

same rows/columns

same border margin

same size of cells
→ great, simplest pipeline

(B) Model is trained offline (e.g., in Colab)

we only need inference in the app

(C) App runs entirely client-side

no backend required

works offline

safe to distribute

9. High-Level Flow in React (Final Architecture)
App.jsx
→ Loads TF model
→ Shows CameraView
CameraView.jsx
→ Captures frames
→ Passes frames to processFrame(frame)
processFrame(frame)
→ run OpenCV pipeline
→ warp perspective
→ extract grid cells
→ convert to tensors
→ run ML model
→ return grid matrix
App.jsx
→ receives updates
→ shows UI visualization
